app/src/main/java/app/musikus/ui/activesession/ActiveSessionViewModel.kt

```diff
@@ -144,40 +136,46 @@ class ActiveSessionViewModel @Inject constructor(
         initialValue = listOf()
     )
 
-    private val currentlyPracticedSection = sessionServiceState.map {
-        it.sections.lastOrNull()
-    }.stateIn(
-        scope = viewModelScope,
-        started = SharingStarted.WhileSubscribed(5000),
-        initialValue = null
-    )
-
-
-    private val totalDurationRoundedDownToNextSecond = sessionServiceState.map { sessionState ->
-        sessionState.sections.sumOf {
-            (it.duration ?: sessionState.currentSectionDuration).inWholeMilliseconds
-        }.milliseconds.inWholeSeconds.seconds
-    }.stateIn(
-        scope = viewModelScope,
-        started = SharingStarted.WhileSubscribed(5000),
-        initialValue = 0.seconds
-    )
 
     /**
      *  ---------------- Composing the Ui state --------------------
      */
 
+
+    private val sessionState = combine(
+        activeSessionUseCases.getCompletedSections(),
+        timeProvider.clock(100.milliseconds)
+    ) { completedSections, _ ->
+        try {
+            SessionViewModelState(
+                sessionDuration = activeSessionUseCases.getPracticeTime(),
+                sections = completedSections,
+                activeSection = activeSessionUseCases.getRunningSection(),
+                ongoingPauseDuration = activeSessionUseCases.getOngoingPauseDuration(),
+                isPaused = activeSessionUseCases.getPausedState(),
+                startTimestamp = activeSessionUseCases.getStartTime()
+            )
+        } catch (e: IllegalStateException) {
+            null
+        }
+    }
+
     private val libraryCardHeaderUiState = combine(
         foldersWithItems,
         _selectedFolderId,
-        currentlyPracticedSection
-    ) { foldersWithItems, selectedFolderId, currentlyPracticedSection ->
+        activeSessionUseCases.getCompletedSections()    // to update the active folder id when a section is clicked
```

Comment:
You mean because I'm piping the `activeSessionUseCases.getCompletedSections()` directly into `combine()`? The useCase directly returns a flow and I don't need the flow elsewhere... Or are there any disadvantages to first mirroring it into a stateflow?

Reply:
the flow is literally used again 20 lines earlier 😅 also we only want to use stateflows in the viewmodel to prevent updates from the other flows in the combine getting lost.